### ISSUE TITLE
fdupes: option to replace duplicates with hard links

### ISSUE THREAD

**Issue by sandrotosi** [adrianlopezroche/fdupes#46]
_From @sandrotosi on December 20, 2015 14:4_

_From [matrixhasu](https://code.google.com/u/matrixhasu/) on October 08, 2009 22:13:05_

Debian bug `#284274` - http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=284274 From: Rupert Levene \<rlevene@uwaterloo.ca>

It would be nice to have the option of telling fdupes to replace
duplicate files with hard links. This would be a more symmetric
behaviour than using symlinks.

\-----

From: Javier Fernández-Sanguino Peña \<jfs@computer.org>

Attached is a patch to the program sources (through the use of a dpatch patch
in the Debian package) that adds a new -L / --linkhard option to fdupes. This
option will replace all duplicate files with hardlinks which is useful in
order to reduce space.

It has been tested only slightly, but the code looks (to me) about right.

**Attachment:** [284274_fdupes_hardlink_repace.diff](http://code.google.com/p/fdupes/issues/detail?id=8)

_Original issue: http://code.google.com/p/fdupes/issues/detail?id=8_

_Copied from original issue: sandrotosi/fdupes-issues#5_

**Comment by sandrotosi** [comment]
_From [mika...@gmail.com](https://code.google.com/u/110951070450373132512/) on July 28, 2010 03:42:27_

There is a typo in the help output in this patch, after the --debug line, "each set of duplicates without prompting the user" is spuriously output again, repeated from the --linkhard output.

**Comment by sandrotosi** [comment]
_From [matrixhasu](https://code.google.com/u/matrixhasu/) on August 01, 2010 06:34:25_

I can see this (it's on the output of 'fdupes --help'); I'd just remove

+  printf("                  \teach set of duplicates without prompting the user\n");

after --debug.

**Comment by sandrotosi** [comment]
_From [zaar...@gmail.com](https://code.google.com/u/116363330358804917839/) on August 26, 2010 10:23:00_

the manpage has typo: the advertised option isn't "--hardlink", but "--linkhard".

**Comment by sandrotosi** [comment]
_From [jean.jor...@gmail.com](https://code.google.com/u/115971232610617876835/) on December 03, 2010 22:06:27_

I ran fdupes with this option. Unfortunately, I forgot that I was on a vfat partition. 
It looks like this basically deleted all files. 
Recovery in progress using photorec.

**Comment by sandrotosi** [comment]
_From [fszcze...@gmail.com](https://code.google.com/u/105694354510325746126/) on October 29, 2011 19:30:22_

I have an independently developed patch to do the same thing. It's a bit more careful about creating the links, so it should fail safely on filesystems that don't support links. It also merges sets of previously hardlinked files correctly (see issue #22 ).

The patch includes some code cleanups and optimizations as well, which I split out separately:

0001-Whitespace-cleanups.patch
- No functional changes, just whitespace changes to improve readability.

0002-Use-strdup-instead-of-malloc-strcpy-pairs.patch
- Use strdup instead of separate malloc/strcpy calls.
- Use calloc instead of individually clearing members
- Avoid multiple passes over strings when concatenating dir and dirinfo->d_name

0003-Cache-stat-2-results.patch
- Cache results from stat(2) calls for improved performance.
- Reorganize grokdir() flow to avoid multiple free/continue blocks.

0004-Add-relink-support.patch
- Rework relink function to start with a temp file in the right directory then rename it to the correct name. This ensures that any likely errors are detected before the file is lost.
- relinkfiles correctly handles duplicate files across multiple filesystems, linking together those files that reside on the same filesystem.
- Update sort to:
  - Prefer the file with more hardlinks
  - Fall back to a filename comparison if link counts and mtimes are the same to ensure stable results.

Relink mode automatically enables --hardlinks, it doesn't make much sense to ignore existing hardlinked files while crating new ones. It also skips empty files, hard links save nothing in this case. If the user really wants to link the empty files, we provide a --relinkempty mode that will do so.

**Attachment:** [0001-Whitespace-cleanups.patch 0002-Use-strdup-instead-of-malloc-strcpy-pairs.patch 0003-Cache-stat-2-results.patch 0004-Add-relink-support.patch](http://code.google.com/p/fdupes/issues/detail?id=8#hc5)

**Comment by sandrotosi** [comment]
_From [sandro.tosi](https://code.google.com/u/sandro.tosi/) on June 17, 2012 12:27:19_

The patch as presented in the original post is bugged, in case some of the files are on 2 different filesystems, as reported on http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=677419 :

\>>>
By mistake, I have been using fdupes -rL on what i thought were
directories, but one was a symbolic link to a directory on another
file system. It found duplicates accross file systems, ... and
removed one end. For the little story, the end where it removed
files was /lib, and I was deduplicating chroots, so it found
plenty of duplicates... try running anything without ld.so...

Anyways, this can easily be reproduced. Assuming /mnt/a and /mnt/b
are two different filesystems:
$ echo a > /mnt/a/a
$ echo a > /mnt/a/b
$ echo a > /mnt/b/a
$ fdupes -rL /mnt/a /mnt/b
  [+] /mnt/a/a
  [h] /mnt/a/b
-- unable to create a hardlink for the file: Invalid cross-device link
  [!] /mnt/a/a
$ ls /mnt/b
$
\<<<

**Comment by xambroz** [comment]
+1 for this feature

**Comment by Harvie** [comment]
I think user should be able to choose between hardlinks and symlinks. Hardlinks make more sense most of the time, but sometimes you might want to use symlinks just to make it more obvious that there is a link.

**Comment by wmertens** [comment]
Is there a reason this is not merged? There are many versions of fdupes out there, and most of them have this functionality, so it is weird that this one doesn't have it.

Furthermore, the incremental flag for the deleting could also apply for hardlinking. Then there would also be no problem with ignoring files linked to the same inode: While traversing, keep track of which inodes you linked to other inodes, and whenever you encounter that same old inode, you link it to the new inode. You can drop the inode from memory as soon as its refcount is 1.

**Comment by aviogit** [comment]
I agree with @wmertens, this feature should be merged, I'm tired of doing deduplication with fslint-gui...

**Comment by jbruchon** [comment]
If you really can't wait for a merge into fdupes master, I've merged it in my [jdupes](https://github.com/jbruchon/jdupes) fork which you are welcome to try. A symlinking target is also present. It's probably a less cryptic command-line option than fslint is.

**Comment by aviogit** [comment]
Thanks, I'll give it a try.

**Comment by adrianlopezroche** [comment]
The reason it isn't merged is the question of what to do should the link
operation fail. The program would be breaking the promise that files will
be linked instead of deleted. Indeed, some files cannot be hardlinked at
all, such as when they lie on different volumes.

**Comment by jbruchon** [comment]
That is a serious problem I ran into with that patch. I decided to rename the original file and on any sort of failure it would be renamed back, being deleted only once the hard link operation succeeds. Unfortunately, that also led to a lot of extra error checking code being added. It's not pretty.

**Comment by aviogit** [comment]
On Linux one can use stat (http://stackoverflow.com/a/8483435/1396334) to look at the inode number and the number of hard links for each file (before and after the hard link operation). As @jbruchon did, the file should only be renamed at first (like rsync does, for example) and if the hardlink fails, it should be safely renamed back to the original name. I know, this will require a lot of testing, but sounds feasible.

**Comment by jbruchon** [comment]
It should be sufficient to receive a return value indicating success from the link call. Checking the inode count would be redundant.

**Comment by wmertens** [comment]
First link a temp name, and if that succeeds, remove the file and rename
the new link. If it fails, keep and use the file as the new inode for
linking identical files.

**Comment by adrianlopezroche** [comment]
There's another problem: making sure not to overwrite an existing file when
the dupe is renamed. Easy with rename2 on Linux, but I can't find a
portable call that's guaranteed not to overwrite an existing file.

**Comment by wmertens** [comment]
Err - the idea is to overwrite the existing file with the new link, no?

**Comment by adrianlopezroche** [comment]
That's the last step. First you move the dupe aside by renaming it. The
choice of new name is critical.

**Comment by wmertens** [comment]
You can create the link anywhere on the same filesystem, for example in a
temporary directory with no other files…

**Comment by jbruchon** [comment]
My solution was to append `.___jdupes___.tmp` or something similar to the temporary name. Any concerns about such a specific extension existing can be resolved by checking for the existence of the temporary name prior to using it.

**Comment by hsoft** [comment]
What's wrong with `ln -f src dest`? If linking fails, `dest` is not touched. For the benefit of people wanting a quick solution, here's a bash script:

```
#!/bin/bash

previous=""
while read -r line; do
    if [[ -n "$previous" ]] && [[ -n "$line" ]]; then
        echo "$line => $previous"
        ln -f "$previous" "$line"
    fi
    previous=$line
done
```

You can then use it with `fdupes <args> <path> | pasted-script.sh`

**Comment by wmertens** [comment]
@hsoft if you're going to do that, you can also just implement the whole thing in bash with md5sum, sort etc.

**Comment by hsoft** [comment]
@wmertens I don't understand what you mean. You mean not using `fdupes`? That would be needless repetition of work. It already takes care of inode checking and the size -> md5 -> byte-to-byte escalation. Why would I want to re-implement that?

**Comment by wmertens** [comment]
Well, why do you want to re-implement this PR? :)

**Comment by hsoft** [comment]
Because it's not merged and it's quicker to use a bash script than to apply patches and recompile fdupes? Don't take it personally, I'm just sharing a solution to people's problems.

I stumbled on this issue by looking for a way to accomplish this. This was the first hit. I provide an immediate and quick solution.

**Comment by wmertens** [comment]
Sorry for coming over as salty. I appreciate your efforts, but it seems like a bit of a hack.

Of course, if this works 100% of the time in all edge cases, maybe it would be better to wrap fdupes in a script that does this. Less code is better code.

**Comment by etique57** [comment]
thanks @hsoft it worked flawlessly in my case.

**Comment by gsakkis** [comment]
Bump; there's a 3+ year old [fork](https://github.com/tobiasschulz/fdupes) that implements this, why is it not merged?

**Comment by jbruchon** [comment]
@gsakkis I can't answer for @adrianlopezroche directly, but I can say that merging that hard link code is a bad decision. [Just to demonstrate why, this is all the stuff my fork has to do to safely replace with links.](https://github.com/jbruchon/jdupes/blob/master/act_linkfiles.c) The code in that fork you linked simply deletes the files and makes hard links to them, but in the event that anything goes wrong with hard linking, you've just deleted your files and that's the end of it.  While the patch would be safe in success cases and in the case of a delete failure, it is not safe in the event of a link failure.

I've had incidents with the code I linked where my paranoia has resulted in not losing files. A good example is if you mistakenly try to hard link on a FAT filesystem which can't take hard links. Oops, your data is gone and there is no way to get it back!

The proper way to do it is to move the target to a temporary file name, do the link, then delete the temporary file name. Any failure = undo each operation performed until you are back at the original state. The worst-case there is that the file gets renamed but can't be renamed back to its original name, but I used a funky extension that's easily lopped off with a shell script or even manually renamed. At least you don't lose the duplicate file paths in that case.

**Comment by gsakkis** [comment]
@jbruchon thanks for the warning; I just found a link to the fork in another issue and didn't pay close attention to it, let alone review the code. I wasn't aware of jdupes; I'll definitely try it out :+1:

**Comment by jbruchon** [comment]
@gsakkis I appreciate it, but it would be helpful if you could do something to help with getting this feature in fdupes, especially since 2.0 brings an ncurses interface that is not found with most other duplicate scanners.

**Comment by wmertens** [comment]
@jbruchon would it not be better to first create a hardlink with a temporary name, and then atomically `rename()`? That will overwrite the original with the hardlink and at no point will the file be missing

**Comment by jbruchon** [comment]
@wmertens You may be right. I didn't put that much thought into it. I solved the loss case that way and didn't try to refine it.

**Comment by Harvie** [comment]
> first create a hardlink with a temporary name, and then atomically rename()

Yes. Correct me if i am wrong, but i believe this is how things are usualy done if you want atomicity on filesystem...

**Comment by jnorthrup** [comment]
I was just about to go and build a bikeshed and thought I would drop in and post this first:

`fdupes -r -n -1  ./|while read; do l=( $REPLY ); for ((i=1;i<${#l[*]};i++)) ; do echo ln -f $l ${l[$i]};done; done`

**Comment by mohag** [comment]
> The reason it isn't merged is the question of what to do should the link operation fail. The program would be breaking the promise that files will be linked instead of deleted. Indeed, some files cannot be hardlinked at all, such as when they lie on different volumes.

Some careful linking should make that exceedingly unlikely: (I propose this (overly careful) process) (basic checks should be done after the steps (e.g. fstat and checking the inodes of the involved names))
* create a temporary hardlink to the original file (that the duplicate would be hardlinked to eventually) in the directory as where the link should go (This checks that hardlinks can be created and will fail e.g. on different filesystems)
* hardlink the file that is to be replaced by a hardlink to a new name (e.g. filename.bak if there is no conflicts) (This is to provide a backup if a later step fails)
* Unlink the old name
* link the source file to the final name
* Unlink the temporary test link (this can happen a bit earlier as well)
* Unlink the backup of the replaced file (the replaced file would only be deleted here and disk space freed here...)

If any any errors occur in this process, exit with a fatal error (with details on where the data can be found and what failed). This should ensure no data loss... (An attempt to restore things to how they were (relink original names, unlink backup / additional names) can be made as well, with a warning issued instead of crashing if it is successfully undone)

Avoiding filename conflicts for two temporary files is one issue (one way is to use .bak and .new and if that conflicts, a hash (md5, since the code is already present) of the filename + a fixed string (or a few random bytes) (that would only conflict if files were specifically created to conflict, which should be impossible if adding a few random bytes to the hash input instead of a fixed string)

(I see atomic renames are also proposed, I was not aware of those and they might simplify the process somewhat....) (I was under the impression that a rename is a link to new name, unlink on old name on most file systems)

One other thing that I wonder about is whether identical files with different metadata should be considered duplicates (when hardlinking, I suspect not, at least by default) (hardlinking unmodified files in user home dirs to those in /etc/skel (with other permissions) would mess with stuff...)

**Comment by jasonkhanlar** [comment]
> I was just about to go and build a bikeshed and thought I would drop in and post this first:
> 
> `fdupes -r -n -1 ./|while read;do l=( $REPLY );for ((i=1;i<${#l[*]};i++));do echo ln -f $l ${l[$i]};done;done`

Tinkering with the idea of automating hardlinking, I encountered an issue that I will try to illustrate, but basically if there are existing hardlinked files and a duplicate file not part of the hardlink is identified by fdupes, depending on the `stat` time of the files, the line above may result in attempting to create a hardlink of the nonhardlinked file to be hardlinked to it, whilst simultaneously breaking the formerly hardlinked file from the hardlink to the other files, therefore creating two sets of hardlinked files with duplicate data. For example:

```
# echo one > 1
# ln -f 1 2
# echo one > 3
# touch 1
# fdupes -r -n -1 ./|while read;do l=( $REPLY );for ((i=1;i<${#l[*]};i++));do echo ln -f $l ${l[$i]};done;done
ln -f ./3 ./1
# stat *
  File: 1
  Size: 4         	Blocks: 8          IO Block: 4096   regular file
Device: 259,2	Inode: 62390579    Links: 2
...
  File: 2
  Size: 4         	Blocks: 8          IO Block: 4096   regular file
Device: 259,2	Inode: 62390579    Links: 2
...
  File: 3
  Size: 4         	Blocks: 8          IO Block: 4096   regular file
Device: 259,2	Inode: 62390583    Links: 1
...
# fdupes -r -n -1 ./|while read;do l=( $REPLY );for ((i=1;i<${#l[*]};i++));do ln -f $l ${l[$i]};done;done
# stat *
  File: 1
  Size: 4         	Blocks: 8          IO Block: 4096   regular file
Device: 259,2	Inode: 62390583    Links: 2
...
  File: 2
  Size: 4         	Blocks: 8          IO Block: 4096   regular file
Device: 259,2	Inode: 62390579    Links: 1
...
  File: 3
  Size: 4         	Blocks: 8          IO Block: 4096   regular file
Device: 259,2	Inode: 62390583    Links: 2
...
# fdupes -r -n -1 ./|while read;do l=( $REPLY );for ((i=1;i<${#l[*]};i++));do echo ln -f $l ${l[$i]};done;done
ln -f ./1 ./2
```

Note the very last two lines show that even after running the script to hardlink the duplicates, the operation merely caused an existing hardlinked file in one group of duplicated data files to be relocated into a different group of hardlinked files with the same data, and resulting in preserving two sets of duplicated data files.

**Comment by jnorthrup** [comment]
> > I was just about to go and build a bikeshed and thought I would drop in and post this first:
> > `fdupes -r -n -1 ./|while read;do l=( $REPLY );for ((i=1;i<${#l[*]};i++));do echo ln -f $l ${l[$i]};done;done`
> 
> Tinkering with the idea of automating hardlinking, I encountered an issue that I will try to illustrate, but basically if there are existing hardlinked files and a duplicate file (not hardlinked to any or the same group) is identified by fdupes that is not part of the hardlinked group, depending on the `stat` time of the files, the line above may result in attempting to create a hardlink of the nonhardlinked file to be hardlinked to the hardlinked file, thus separating the formerly hardlinked file from the group to be hardlinked into another group of duplicate data files, therefore creating two sets of hardlinked files with duplicate data. For example:
> 
> ```
> # echo one > 1
> # ln -f 1 2
> # echo one > 3
> # touch 1
> # fdupes -r -n -1 ./|while read;do l=( $REPLY );for ((i=1;i<${#l[*]};i++));do echo ln -f $l ${l[$i]};done;done
> ln -f ./3 ./1
> # stat *
>   File: 1
>   Size: 4         	Blocks: 8          IO Block: 4096   regular file
> Device: 259,2	Inode: 62390579    Links: 2
> ...
>   File: 2
>   Size: 4         	Blocks: 8          IO Block: 4096   regular file
> Device: 259,2	Inode: 62390579    Links: 2
> ...
>   File: 3
>   Size: 4         	Blocks: 8          IO Block: 4096   regular file
> Device: 259,2	Inode: 62390583    Links: 1
> ...
> # fdupes -r -n -1 ./|while read;do l=( $REPLY );for ((i=1;i<${#l[*]};i++));do ln -f $l ${l[$i]};done;done
> # stat *
>   File: 1
>   Size: 4         	Blocks: 8          IO Block: 4096   regular file
> Device: 259,2	Inode: 62390583    Links: 2
> ...
>   File: 2
>   Size: 4         	Blocks: 8          IO Block: 4096   regular file
> Device: 259,2	Inode: 62390579    Links: 1
> ...
>   File: 3
>   Size: 4         	Blocks: 8          IO Block: 4096   regular file
> Device: 259,2	Inode: 62390583    Links: 2
> ...
> # fdupes -r -n -1 ./|while read;do l=( $REPLY );for ((i=1;i<${#l[*]};i++));do echo ln -f $l ${l[$i]};done;done
> ln -f ./1 ./2
> ```
> 
> Note the very last two lines show that even after running the script to hardlink the duplicates, the operation merely caused an existing hardlinked file in one group of duplicated data files to be relocated into a different group of hardlinked files with the same data, and resulting in preserving two sets of duplicated data files.

im not going to guess at the kernel version and vfs driver variables above.

i recall sharing this relative to carrying Steam library to and from ntfs < - >  btrfs | xfs and the game elite-dangerous at the time would be duplicating 20 gigs of hashed content directories where we could do something about it on linux
